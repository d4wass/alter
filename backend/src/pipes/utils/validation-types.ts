import { ValidationError } from '@nestjs/common';

export type ConvertedValidationError = {
  property: string;
  constraints: any;
  children?: ValidationError[];
};

export type ConstructorTypes =
  | StringConstructor
  | BooleanConstructor
  | NumberConstructor
  | ArrayConstructor
  | ObjectConstructor;
