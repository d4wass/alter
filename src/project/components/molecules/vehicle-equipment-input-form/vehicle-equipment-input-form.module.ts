import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { VehicleEquipmentInputFormComponent } from './vehicle-equipment-input-form.component';
import { MatInputModule } from '@angular/material/input';
import { MatButtonModule } from '@angular/material/button';
import { MatListModule } from '@angular/material/list';
import { MatTooltipModule } from '@angular/material/tooltip';
import { VehicleEquipmentItemModule } from '../../atoms/vehicle-equipment-item/vehicle-equipment-item.module';

@NgModule({
  declarations: [VehicleEquipmentInputFormComponent],
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MatInputModule,
    MatButtonModule,
    MatTooltipModule,
    MatListModule,
    VehicleEquipmentItemModule
  ],
  exports: [VehicleEquipmentInputFormComponent]
})
export class VehicleEquipmentInputFormModule {}
