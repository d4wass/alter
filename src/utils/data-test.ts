export function dataTest(selector: string) {
  return `[data-test="${selector}"]`;
}
