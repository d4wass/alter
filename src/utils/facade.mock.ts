import { Observable, BehaviorSubject } from 'rxjs';

export type FacadeMock<T> = {
  [P in keyof T]: T[P] extends Observable<infer U>
    ? BehaviorSubject<U>
    : T[P] extends (args: infer A) => Observable<infer R>
    ? (args: A) => BehaviorSubject<R>
    : T[P];
};
