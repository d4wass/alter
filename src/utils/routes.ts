export interface urlItem {
  description: string;
  url: string;
}

export const urlList: urlItem[] = [
  { description: 'Learn more', url: 'faq' },
  { description: 'Carculator', url: 'carculator' },
  { description: 'Become a host', url: 'host' },
  { description: 'Create Account', url: 'new-user' },
  { description: 'Your Cars', url: 'vehicles' }
];
