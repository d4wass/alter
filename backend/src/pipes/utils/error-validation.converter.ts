import { ValidationError } from 'class-validator';
import { ConvertedValidationError } from './validation-types';

export class ValidationErrorConverter {
  private _convertedErrors: Array<ConvertedValidationError> = [];
  public get convertedErrors(): Array<ConvertedValidationError> {
    return this._convertedErrors;
  }
  public set convertedErrors(value: Array<ConvertedValidationError>) {
    this._convertedErrors = value;
  }
  private _errors: ValidationError[];

  constructor(errors: ValidationError[]) {
    this._errors = errors;
    this.errorsConverter(this._errors);
  }

  errorsConverter(errors: ValidationError[]): void {
    errors.forEach((error) => {
      if (!error.children.length) {
        this._convertedErrors.push({ property: error.property, constraints: error.constraints });
      } else {
        this.errorsConverter(error.children);
      }
    });
  }
}
