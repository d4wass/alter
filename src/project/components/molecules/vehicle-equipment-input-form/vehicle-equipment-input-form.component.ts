import {
  ChangeDetectionStrategy,
  Component,
  ElementRef,
  EventEmitter,
  Output,
  ViewChild
} from '@angular/core';
import { Validators } from '@angular/forms';
import { MatTooltip } from '@angular/material/tooltip';
import { FormControl } from '@ngneat/reactive-forms';
import { EquipmentValidationMsg } from 'src/+state/models/user.model';
import { isNonNull } from '../../../../utils/helpers';

@Component({
  selector: 'app-vehicle-equipment-input-form',
  template: `
    <div>
      <form>
        <mat-form-field>
          <mat-label>Enter additional equipment</mat-label>
          <input matInput [formControl]="control" #equipmentInput data-test="input" />
        </mat-form-field>
        <div #tooltip="matTooltip" matTooltip matTooltipPosition="right" data-test="tooltip"></div>
        <button
          class="add-btn"
          mat-flat-button
          color="primary"
          (click)="handleAdd($event)"
          data-test="add-btn"
        >
          Add
        </button>
      </form>
      <mat-list>
        <app-vehicle-equipment-item
          *ngFor="let item of equipment; let i = index"
          [equipment]="item"
          (handleClose)="handleRemove(i)"
        ></app-vehicle-equipment-item>
      </mat-list>
    </div>
  `,
  styleUrls: ['./vehicle-equipment-input-form.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  standalone: false
})
export class VehicleEquipmentInputFormComponent {
  @ViewChild('equipmentInput') inputElement!: ElementRef;
  @ViewChild('tooltip') tooltip!: MatTooltip;
  @Output() vehicleEquipment = new EventEmitter<string[]>();

  private whiteSpacesRegex = new RegExp(/^(\s+\S+\s*)*(?!\s).*$/);

  control: FormControl<string> = new FormControl('', [
    Validators.minLength(3),
    Validators.pattern(this.whiteSpacesRegex)
  ]);
  equipment: string[] = [];

  handleAdd(e: Event): void {
    e.preventDefault();

    if (isNonNull(this.control.value) && this.control.valid && this.control.value.length !== 0) {
      this.equipment = this.checkListForRepeatsValue(this.control.value);
      this.control.setValue('');
      this.vehicleEquipment.emit(this.equipment);
    } else {
      this.handleEquipmentValidation(this.control.value);
    }

    this.inputElement.nativeElement.focus();
  }

  handleRemove(index: number): void {
    this.equipment.splice(index, 1);
    this.vehicleEquipment.emit(this.equipment);
  }

  private checkListForRepeatsValue(value: string): string[] {
    const trimedValue = value.trim();
    let vehicleEquipment = [...this.equipment, trimedValue];

    if (this.equipment.includes(trimedValue)) {
      vehicleEquipment = this.equipment;
      this.handleEquipmentValidation(trimedValue);
    } else {
      vehicleEquipment = [...this.equipment, trimedValue];
    }

    return vehicleEquipment;
  }

  private handleEquipmentValidation(value: string): void {
    const validationResult = {
      emptyValue: value.length === 0,
      whiteSpacesValue: !this.whiteSpacesRegex.test(value),
      alreadyAddedValue: this.equipment.includes(value.trim()),
      tooShortValue: value.length <= 3 && value.length !== 0
    };

    this.tooltip.message = this.convertEquipmentValidationMsg(validationResult);
    this.tooltip.show();
  }

  private convertEquipmentValidationMsg(validationResult: EquipmentValidationMsg): string {
    const { emptyValue, whiteSpacesValue, alreadyAddedValue, tooShortValue } = validationResult;
    if (emptyValue) {
      return "You don't provide any value. Please enter equipement name";
    }
    if (whiteSpacesValue) {
      return 'You provide incorrect value. Please enter equipement name';
    }
    if (alreadyAddedValue) {
      return 'You already add this value. Make sure your value is unique';
    }
    if (tooShortValue) {
      return 'Provided value is too short.';
    } else {
      return '';
    }
  }
}
