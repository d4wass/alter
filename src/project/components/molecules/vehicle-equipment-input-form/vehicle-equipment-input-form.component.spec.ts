import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { MatTooltipModule } from '@angular/material/tooltip';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { dataTest } from '../../../../utils/data-test';
import { VehicleEquipmentInputFormComponent } from './vehicle-equipment-input-form.component';

describe('VehicleFeatureCheckboxInputFormComponent', () => {
  let component: VehicleEquipmentInputFormComponent;
  let spectator: Spectator<VehicleEquipmentInputFormComponent>;

  const createComponent = createComponentFactory({
    component: VehicleEquipmentInputFormComponent,
    imports: [ReactiveFormsModule, MatTooltipModule],
    schemas: [NO_ERRORS_SCHEMA],
    shallow: true
  });

  const selectors = {
    addBtn: () => spectator.query(dataTest('add-btn')),
    input: () => spectator.query(dataTest('input')),
    tooltip: () => spectator.query(dataTest('tooltip'))
  };

  beforeEach(() => {
    spectator = createComponent();
    component = spectator.component;
  });

  it('should match to snapshot', () => {
    expect(spectator.fixture).toMatchSnapshot();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('handleAdd()', () => {
    const event = new MouseEvent('click');

    it('should call handle add on btn click', () => {
      const spy = jest.spyOn(component, 'handleAdd');
      spectator.click(selectors.addBtn()!);

      spectator.detectChanges();
      expect(spy).toBeCalledWith(event);
    });

    it('should emit equipment array if control value is valid', () => {
      let result;
      const value = 'controlValue';

      component.control.setValue(value);
      spectator.detectComponentChanges();

      expect(component.control.value).toEqual(value);
      spectator.output('vehicleEquipment').subscribe((x) => (result = x));
      component.handleAdd(event);

      expect(component.control.value).toEqual('');
      expect(result).toEqual([value]);
    });

    it('should emit equipment array with only unique values', () => {
      const equipement = ['value1', 'value2'];
      const value = 'value1';

      component.equipment = equipement;
      component.control.setValue(value);
      spectator.detectComponentChanges();
      component.handleAdd(event);

      expect(component.equipment).toEqual(equipement);
    });

    it('should not emit vehicleEquipment if control value is invalid', () => {
      let result;
      const spy = jest.spyOn(component.vehicleEquipment, 'emit');

      component.control.setValue('');
      component.handleAdd(event);
      spectator.output('vehicleEquipment').subscribe((x) => (result = x));

      expect(result).toBeUndefined();
      expect(spy).not.toBeCalled();
    });

    it('should focus on input element on handleAdd call', () => {
      component.handleAdd(event);
      expect(selectors.input()).toBeFocused();
    });
  });
  describe('handleRemove()', () => {
    it('should emit vehicle equipment array without removed item', () => {
      const spy = jest.spyOn(component.vehicleEquipment, 'emit');
      component.equipment = ['value1', 'value2'];
      component.handleRemove(0);

      expect(spy).toBeCalledWith(['value2']);
    });
  });
  //TODO: create proper cdk-overlay unit test
  describe('tooltip', () => {
    it.each([
      ['', "You don't provide any value. Please enter equipement name"],
      ['   ', 'You provide incorrect value. Please enter equipement name'],
      ['value', 'You already add this value. Make sure your value is unique'],
      ['va', 'Provided value is too short']
    ])('should show tooltip for invalid values', (controlValue, tooltipMsg) => {
      const event = new MouseEvent('click');

      component.equipment = ['value'];
      component.control.setValue(controlValue);
      component.handleAdd(event);

      expect(selectors.tooltip).toBeTruthy();
    });
  });
});
