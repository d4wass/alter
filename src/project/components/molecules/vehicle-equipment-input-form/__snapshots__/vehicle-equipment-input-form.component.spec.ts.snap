// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleFeatureCheckboxInputFormComponent should match to snapshot 1`] = `
<app-vehicle-equipment-input-form
  control={[Function FormControl]}
  equipment={[Function Array]}
  inputElement={[Function ElementRef]}
  tooltip={[Function MatTooltip]}
  vehicleEquipment={[Function EventEmitter_]}
  whiteSpacesRegex={[Function RegExp]}
>
  <div>
    <form
      novalidate=""
    >
      <mat-form-field>
        <mat-label>
          Enter additional equipment
        </mat-label>
        <input
          class="ng-untouched ng-pristine ng-valid"
          data-test="input"
          matinput=""
        />
      </mat-form-field>
      <div
        class="mat-tooltip-trigger"
        data-test="tooltip"
        mattooltip=""
        mattooltipposition="right"
      />
      <button
        class="add-btn"
        color="primary"
        data-test="add-btn"
        mat-flat-button=""
      >
         Add 
      </button>
    </form>
    <mat-list>
    </mat-list>
  </div>
</app-vehicle-equipment-input-form>
`;
